DIPs/DIP1xxx-sw.md

```diff
@@ -0,0 +1,954 @@
+
+### Simple
+
+```d
+struct totalArgs {
+    int tax;
+    int discount;
+}
+
+int total(int subtotal, totalArgs args = totalArgs.init) {
+    return subtotal + args.tax - args.discount;
+}
+
+unittest {
+    assert(42.total == 42);
+    assert(42.total(totalArgs(tax: 50)) == 92);
+    assert(total(42, totalArgs(discount: 20, tax: 50)) == 72);
+
+    int defaultTotal(int subtotal) {
+        return total(subtotal, totalArgs(tax: 20));
+    }
+}
+```
+
+### Phobos
+
+Many functions in [std.process](https://dlang.org/phobos/std_process.html) provide
+many overload to workaround the problem of not being able to just specify a argument
+that comes at the end. For example, let's pick [`spawnProcess`](https://dlang.org/phobos/std_process.html#.spawnProcess)
+and assume the user wants to change the working directory of the spawned process:
+
+```d
+import std.process;
+auto pid = spawnProcess(["dmd", "myprog.d"],
+                        std.stdio.stdin,
+                        std.stdio.stdin,
+                        std.stdio.stderr,
+                        null,
+                        Config.none,
+                        "/my/custom/working/dir, // <- parameter that the user is interested in changing
+);
+```
+
+With in-place struct initialization, it could look like this:
+
+```d
+auto pid = spawnProcess(["dmd", "myprog.d"], SpawnParams(workingdir: "/my/custom/workingdir"));
+```
+
+### Ranges
+
+The syntax can be used analogous to [Tuples](https://dlang.org/phobos/std_typecons.html#.Tuple)
+to conveniently use lambdas within range constructs:
+
+```d
+auto queueCreateInfo = priorities
+	.enumerate
+	.filter!(x => x[1].length > 0)
+	.map!(x => S({c: 10}))
+	.array;
+```
+
+### Plotting
+
+As a specific use-case area, we will have a look at plotting systems.
+Two of the most most popular open source plotting libraries are
+[ggplot](http://ggplot2.org) and [matplotlib](https://matplotlib.org/)
+and they allow to configure plotting in a variety of ways.
+
+#### Python/matplotlib
+
+For example, the rudimentary [plot](https://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.plot)
+API allows a plethora of optional parameters:
+
+```py
+plot(x, y, color='green', linestyle='dashed', marker='o',
+     markerfacecolor='blue', markersize=12).
+```
+
+Currently, a call to theoretical wrapper to Matplotlib would look roughly like this:
+
+```d
+PlotConfig config = {
+    color: 'green',
+    linestyle: 'dashed',
+    marker: 'o',
+    markerfacecolor: 'blue'
+    markersize: 12,
+};
+plot(x, y, config);
+```
+
+#### ggplot/R vs. ggplotd
+
+R supports named arguments, so e.g. [`aes`](http://ggplot2.tidyverse.org/reference/aes.html) can be constructed directly:
+
+```R
+aes(x = mpg, y = wt)
+```
+
+[ggplotd](https://github.com/BlackEdder/ggplotd) currently uses template arguments
+and a lot of compile-time introspection power to allow arbitrary order of the name of parameters:
+
+```d
+aes!("x", "y")(dat1.value1, dat1.value2)
+```
+
+#### Vulkan
```

Comment:
This heading and the next two should use `###` for level-3, they're not part of *Plotting*.